### ISSUE TITLE
Achievement Mine Sweeper

### ISSUE THREAD

**Issue by Asterc** [TrinityCore/TrinityCore#15938]
Achievement [Mine Sweeper](http://wotlk.openwow.com/achievement=1428) doesn't work.
"Get caught in 10 consecutive land mine explosions in the Sparksocket Minefield without landing."
[Criteria spell](http://wotlk.openwow.com/spell=57064)

hash f8f815dbc3a9
branch 3.3.5

**Comment by w5860363** [comment]
Known issues Mine Sweeper (1428) still dont work https://github.com/TrinityCore/TrinityCore/issues/965

**Comment by w5860363** [comment]
https://raw.githubusercontent.com/M0ra/Core/af53dc3687253e8c9a4e2831b99ac2569de97aff/src/server/scripts/Spells/spell_generic.cpp

Perhaps GIT's hacking script can solve this problem.

**Comment by w5860363** [comment]
We need a spell check 

Aura const* aura = source->GetAura(57099);
        if (!(aura && aura->GetStackAmount() == 10))
            return false;

**Comment by DevRival** [comment]
Я сделал данное достижение , но уже сомневаюсь делать слияние . Есть маленькие ошибки.
[video](https://youtu.be/CKRFkhzFxgc)

**Comment by tkrokli** [comment]
translate.google.com suggests this translation of your Russian words:
"I made this achievement, but I doubt do merge. There are small errors."

Could you provide a more precise translation of "I doubt do merge" ? :)
I am guessing that it has got something to do with not doing 10 explosions in a row,
only 2 to get the achievement (Салер?).

**Comment by DevRival** [comment]
You're right, I do not know much English, but that does not prevent use of the translator. I added three scripts to achieve spells and mines. And noticed a small mistake, achieving counted 10 bombings, even if you fall.

**Comment by tkrokli** [comment]
> **ccrs** commented on 13 Oct (2015-10-13 20:40 CEST):

> open a new issue with this pls https://github.com/TrinityCore/TrinityCore/issues/965#issuecomment-4863094

OK, here it is. All of the code below is written by Vincent-Michael  on 1 Apr 2012. I have only corrected a typo in `SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT` and placed the script at the end of /src/server/scripts/Spells/spell_generic.cpp in a new branch based on commit ab83150 (line numbers different from 2012):

```diff
diff --git a/sql/updates/world/2015_12_31_99_world.sql b/sql/updates/world/2015_12_31_99_world.sql
new file mode 100644
index 0000000..3483b18
--- /dev/null
+++ b/sql/updates/world/2015_12_31_99_world.sql
@@ -0,0 +1,5 @@
+-- 
+-- Landmine Knockback Achievement Aura, spell ID 57099;
+DELETE FROM `spell_script_names` WHERE `spell_id` = 57099;
+INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
+(57099,'spell_gen_landmine_knockback_achievement_aura');
```
```diff
diff --git a/src/server/scripts/Spells/spell_generic.cpp b/src/server/scripts/Spells/spell_generic.cpp
index ca8c9a9..6e7296c 100644
--- a/src/server/scripts/Spells/spell_generic.cpp
+++ b/src/server/scripts/Spells/spell_generic.cpp
@@ -4120,6 +4120,51 @@ public:
     }
 };
 
+enum LandmineKnockbackAchievement
+{
+    SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT  = 57064
+};
+
+class spell_gen_landmine_knockback_achievement_aura : public SpellScriptLoader
+{
+    public:
+        spell_gen_landmine_knockback_achievement_aura() : SpellScriptLoader("spell_gen_landmine_knockback_achievement_aura") { }
+
+        class spell_gen_landmine_knockback_achievement_aura_SpellScript : public SpellScript
+        {
+            PrepareSpellScript(spell_gen_landmine_knockback_achievement_aura_SpellScript);
+
+            bool Validate(SpellInfo const* /*spell*/)
+            {
+                if (!sSpellMgr->GetSpellInfo(SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT))
+                    return false;
+                return true;
+            }
+
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (Player* target = GetHitPlayer())
+                {
+                    Aura const* aura = target->GetAura(GetSpellInfo()->Id);
+                    if (!(aura && aura->GetStackAmount() == 10))
+                        return;
+
+                    target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
+                }
+            }
+
+            void Register()
+            {
+                OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_achievement_aura_SpellScript::HandleScript, EFFECT_1, SPELL_EFFECT_APPLY_AURA);
+            }
+        };
+
+        SpellScript* GetSpellScript() const
+        {
+            return new spell_gen_landmine_knockback_achievement_aura_SpellScript();
+        }
+};
+
 void AddSC_generic_spell_scripts()
 {
     new spell_gen_absorb0_hitlimit1();
@@ -4205,4 +4250,5 @@ void AddSC_generic_spell_scripts()
     new spell_gen_gm_freeze();
     new spell_gen_stand();
     new spell_gen_mixology_bonus();
+    new spell_gen_landmine_knockback_achievement_aura();
 }
```
Just like Vincent-Michael wrote in his post, I have not tested this yet myself, but I am going to test it.

Also in #965 , treak suggested adding 4 lines to /src/server/game/Spells/SpellMgr.cpp like this:
```diff
diff --git a/src/server/game/Spells/SpellMgr.cpp b/src/server/game/Spells/SpellMgr.cpp
index ec616c1..4611c1c 100644
--- a/src/server/game/Spells/SpellMgr.cpp
+++ b/src/server/game/Spells/SpellMgr.cpp
@@ -3338,6 +3338,10 @@ void SpellMgr::LoadSpellInfoCorrections()
                 //! HACK: This spell break quest complete for alliance and on retail not used °_O
                 spellInfo->Effects[EFFECT_0].Effect = 0;
                 break;
+            case 54402: // Detonation - Mine Sweeper Achievement
+                spellInfo->EffectImplicitTargetB[EFFECT_0] = TARGET_UNIT_DEST_AREA_ALLY;
+                spellInfo->EffectImplicitTargetB[EFFECT_1] = TARGET_UNIT_DEST_AREA_ALLY;
+                break;
             // VIOLET HOLD SPELLS
             //
             case 54258: // Water Globule (Ichoron)
```
I have not tested any of this yet, but you are of course free to test this and give feedback.

**Comment by DevRival** [comment]
Yes there is a script , but the script has two mines and achievements.

**Comment by w5860363** [comment]
I've seen similar code from GIThub.
https://github.com/M0ra/Core/blob/master/src/server/scripts/Spells/spell_generic.cpp

**Comment by tkrokli** [comment]
Yes, but that is from 7 months ago and the user Alex / M0ra does not follow TrinityCore standards.
There are at least 2 sections removed in the spell_generic.cpp file on that user's repo, compared to TC.

**Comment by tkrokli** [comment]
This code suggested by @treak does not compile:
```diff
diff --git a/src/server/game/Spells/SpellMgr.cpp b/src/server/game/Spells/SpellMgr.cpp
index ec616c1..4611c1c 100644
--- a/src/server/game/Spells/SpellMgr.cpp
+++ b/src/server/game/Spells/SpellMgr.cpp
@@ -3338,6 +3338,10 @@ void SpellMgr::LoadSpellInfoCorrections()
                 //! HACK: This spell break quest complete for alliance and on retail not used °_O
                 spellInfo->Effects[EFFECT_0].Effect = 0;
                 break;
+            case 54402: // Detonation - Mine Sweeper Achievement
+                spellInfo->EffectImplicitTargetB[EFFECT_0] = TARGET_UNIT_DEST_AREA_ALLY;
+                spellInfo->EffectImplicitTargetB[EFFECT_1] = TARGET_UNIT_DEST_AREA_ALLY;
+                break;
             // VIOLET HOLD SPELLS
             //
             case 54258: // Water Globule (Ichoron)
```
I get the following errors:
```
9>\TrinityCore\src\server\game\Spells\SpellMgr.cpp(3342): error C2039: 'EffectImplicitTargetB' : is not a member of 'SpellInfo'
9> \TrinityCore\src\server\game\spells\SpellInfo.h(294) : see declaration of 'SpellInfo'
9>\TrinityCore\src\server\game\Spells\SpellMgr.cpp(3343): error C2039: 'EffectImplicitTargetB' : is not a member of 'SpellInfo'
9> \TrinityCore\src\server\game\spells\SpellInfo.h(294) : see declaration of 'SpellInfo'
```
... so I will test compiling without the code in SpellMgr.cpp instead (only using Vincent-Michael's code).

**Comment by Demonid** [comment]
@tkrokli 

Use this:

```diff
            case 54402: // Detonation - Mine Sweeper Achievement
                spellInfo->Effects[EFFECT_0].TargetB = SpellImplicitTargetInfo(TARGET_UNIT_DEST_AREA_ALLY);
                spellInfo->Effects[EFFECT_1].TargetB = SpellImplicitTargetInfo(TARGET_UNIT_DEST_AREA_ALLY);
                break;

**Comment by tkrokli** [comment]
OK, thanks, I will test that. Vincent-Michael's code compiles, but does not give any achievement.

**Comment by Re3os** [comment]
@tkrokli 
``` C++
41742f8046dee2a3d46826b90a737fea744885e9
 sql/DR_update/world_04.12.2015.sql          | 10 ++++++-
 src/server/scripts/Spells/spell_generic.cpp | 42 ++++++++++++++++++++++++++++-
 2 files changed, 50 insertions(+), 2 deletions(-)

diff --git a/sql/DR_update/world_04.12.2015.sql b/sql/DR_update/world_04.12.2015.sql
index 56ed488..7b3e382 100644
--- a/sql/DR_update/world_04.12.2015.sql
+++ b/sql/DR_update/world_04.12.2015.sql
@@ -6,5 +6,13 @@ INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `type`, `comm
 (10278, -46924, 1, 'Hand of protection (Rank3) - remove bladestorm'),
 (66009, -46924, 1, 'Hand of protection (rank1 dupe) - remove bladestorm');
 
+-- Mount Leon
+UPDATE `creature_template` SET `modelid1` = 31047 WHERE (entry = 11195);
 
-UPDATE `creature_template` SET `modelid1` = 31047 WHERE (entry = 11195);
\ No newline at end of file
+-- Fix achivment Mine Sweeper
+UPDATE `creature_template` SET `flags_extra` = '130' WHERE `entry` = '29397'; 
+DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = '54355';
+INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `comment`) VALUES ('54355', '54402', 'Landmine Knockback'); 
+DELETE FROM `disables` WHERE `sourceType` = '4' AND `entry` = '5258'; 
+DELETE FROM `spell_script_names` WHERE `spell_id` = '57099';
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES ('57099', 'spell_gen_landmine_knockback'); 
diff --git a/src/server/scripts/Spells/spell_generic.cpp b/src/server/scripts/Spells/spell_generic.cpp
index 23dadf2..ac9dbe6 100644
--- a/src/server/scripts/Spells/spell_generic.cpp
+++ b/src/server/scripts/Spells/spell_generic.cpp
@@ -3503,7 +3503,8 @@ class spell_gen_upper_deck_create_foam_sword : public SpellScriptLoader
 
 enum VehicleScaling
 {
-    SPELL_GEAR_SCALING      = 66668
+    SPELL_GEAR_SCALING      = 66668,
+    SPELL_SOTA_DEMO_SCALING = 65636
 };
 
 class spell_gen_vehicle_scaling : public SpellScriptLoader
@@ -3533,6 +3534,10 @@ class spell_gen_vehicle_scaling : public SpellScriptLoader
                         factor = 1.0f;
                         baseItemLevel = 205;
                         break;
+                    case SPELL_SOTA_DEMO_SCALING:
+                        factor = 1.0f;
+                        baseItemLevel = 230; // based of ~185k hp for ~270 item lvl
+                        break;
                     default:
                         factor = 1.0f;
                         baseItemLevel = 170;
@@ -4138,6 +4143,40 @@ public:
     }
 };
 
+class spell_gen_landmine_knockback : public SpellScriptLoader
+{
+public:
+    spell_gen_landmine_knockback() : SpellScriptLoader("spell_gen_landmine_knockback") { }
+
+    class spell_gen_landmine_knockback_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_gen_landmine_knockback_SpellScript);
+
+        void HandleScript(SpellEffIndex /*effIndex*/)
+        {
+            if (Player* target = GetHitPlayer())
+            {
+                Aura const* aura = GetHitAura();
+                if (!aura || aura->GetStackAmount() != 10)
+                    return;
+
+                AchievementEntry const* achiev = sAchievementStore.LookupEntry(1428);
+                target->CompletedAchievement(achiev);
+            }
+        }
+
+        void Register() override
+        {
+            OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_SpellScript::HandleScript, EFFECT_1, SPELL_EFFECT_APPLY_AURA);
+        }
+    };
+
+    SpellScript* GetSpellScript() const override
+    {
+        return new spell_gen_landmine_knockback_SpellScript();
+    }
+};
+
 void AddSC_generic_spell_scripts()
 {
     new spell_gen_absorb0_hitlimit1();
@@ -4223,4 +4262,5 @@ void AddSC_generic_spell_scripts()
     new spell_gen_gm_freeze();
     new spell_gen_stand();
     new spell_gen_mixology_bonus();
+	new spell_gen_landmine_knockback();
 }
````

**Comment by tkrokli** [comment]
OK, just to get an overview of the changes, let's take a look at your SQL:
  1. change the `flags_extra` value on http://www.wowhead.com/npc=29397/land-mine-bunny from 128 `CREATURE_FLAG_EXTRA_TRIGGER` (creature is trigger-NPC (invisible to players only)) to 130, include `CREATURE_FLAG_EXTRA_CIVILIAN` (creature does not aggro (ignore faction/reputation hostility))
  2. insert `spell_trigger` http://www.wowhead.com/spell=54355/detonation and `spell_effect` http://www.wowhead.com/spell=54402/land-mine-knockback into the table `spell_linked_spell`
  3. remove `DISABLE_TYPE_ACHIEVEMENT_CRITERIA`(4), 5258 (entry) (name?) from `disables`
  4. insert spell 57099, http://www.wowhead.com/spell=57099/landmine-knockback-achievement-aura, into `spell_script_names` to enable `spell_gen_landmine_knockback` in `spell_generic.cpp`

Will test your diff and report back. The TC devs might disapprove of changing `flags_extra`, though.

**Comment by Re3os** [comment]
@tkrokli entry 5258,sory it's my fail.More updates one file.

**Comment by tkrokli** [comment]
TrinityCore rev. 6e226ec4f398 2015-12-09 04:28:46 +0100 (landmine_knockback_achievement branch) (Win64, Release) - based on TC core rev. cc685ac and TDB 335.60 / 2015_12_08_02_world.sql
 
![wowscrnshot_120915_080326](https://cloud.githubusercontent.com/assets/4903600/11678867/fc57e9be-9e4b-11e5-88fe-4d0ce0fc11ac.jpg)

Well, it seems to work, even though I'm such a slow player that I needed to use the lvl 77 "Loaner" trick to be able to hit the mines fast enough in succession. I am sure someone with better gamer skills than me could do it with the character running on its feet. Anyway, this is my current diff and SQL:
```sql
UPDATE `creature_template` SET `flags_extra`= 130 WHERE `entry`= 29397;

DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = 54355;
INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `comment`) VALUES
(54355, 54402, 'Landmine Knockback');

DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_gen_landmine_knockback';
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
(57099, 'spell_gen_landmine_knockback');
```
```diff
diff --git a/src/server/scripts/Spells/spell_generic.cpp b/src/server/scripts/Spells/spell_generic.cpp
index ca8c9a9..22cc3f4 100644
--- a/src/server/scripts/Spells/spell_generic.cpp
+++ b/src/server/scripts/Spells/spell_generic.cpp
@@ -4120,6 +4120,40 @@ public:
     }
 };
 
+class spell_gen_landmine_knockback : public SpellScriptLoader
+{
+public:
+    spell_gen_landmine_knockback() : SpellScriptLoader("spell_gen_landmine_knockback") { }
+
+    class spell_gen_landmine_knockback_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_gen_landmine_knockback_SpellScript);
+
+        void HandleScript(SpellEffIndex /*effIndex*/)
+        {
+            if (Player* target = GetHitPlayer())
+            {
+                Aura const* aura = GetHitAura();
+                if (!aura || aura->GetStackAmount() != 10)
+                    return;
+
+                AchievementEntry const* achiev = sAchievementStore.LookupEntry(1428);
+                target->CompletedAchievement(achiev);
+            }
+        }
+
+        void Register() override
+        {
+            OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_SpellScript::HandleScript, EFFECT_1, SPELL_EFFECT_APPLY_AURA);
+        }
+    };
+
+    SpellScript* GetSpellScript() const override
+    {
+        return new spell_gen_landmine_knockback_SpellScript();
+    }
+};
+
 void AddSC_generic_spell_scripts()
 {
     new spell_gen_absorb0_hitlimit1();
@@ -4205,4 +4239,5 @@ void AddSC_generic_spell_scripts()
     new spell_gen_gm_freeze();
     new spell_gen_stand();
     new spell_gen_mixology_bonus();
+    new spell_gen_landmine_knockback();
 }
```
The real question now is if this is classified as hack or not (I don't know, need a TC dev to verify).

**Comment by Re3os** [comment]
@tkrokli It's my fail,sql include more DB update.

**Comment by tkrokli** [comment]
Hmm? Do you mean `DELETE FROM `disables` WHERE `sourceType` = '4' AND `entry` = '5258';`
or something else? I already removed it because I could not find any connection with this issue.
But are you trying to say that you meant to include even more than you have written in your diff ?

**Comment by Re3os** [comment]
@tkrokli it's my custon change on sql update.

**Comment by tkrokli** [comment]
Already done before my testing. See my SQL above (also copy here):
```sql
UPDATE `creature_template` SET `flags_extra`= 130 WHERE `entry`= 29397;

DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = 54355;
INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `comment`) VALUES
(54355, 54402, 'Landmine Knockback');

DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_gen_landmine_knockback';
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
(57099, 'spell_gen_landmine_knockback');
```

**Comment by Killyana** [comment]
If this works correctly and without any hack, open a PR.

**Comment by tkrokli** [comment]
Well, I was hoping someone, maybe @ccrs or @jackpoz, could say more about the code I tested, it was added to straigth TC 3.3.5 source on commit  cc685ac and TDB 335.60 / 2015_12_08_02_world.sql .
I would feel better about opening a PR on it if someone could confirm if the diff I posted is not a hack.

**Comment by ccrs** [comment]
meh, sniffs are the answer.
Probably there is a spell handling the achiev reward, if not this is all correct imho.

**Comment by tkrokli** [comment]
OK, I see that the achievement is still available in 6.2 (latest comment on wowhead is from 6.2.2).
Hope to hear from someone with access to a sniff on this achievement soon(™) .

**Comment by Killyana** [comment]
The spell is http://fr.wowhead.com/spell=57064/landmine-knockback-achievement it's already used on the script.

**Comment by ccrs** [comment]
not in last @tkrokli version, u need to use
````c++
target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
````

**Comment by tkrokli** [comment]
Thanks for the comment, I was thinking the same that it was not part of my last diff.
Thanks also for the extra code input. :+1:

**Comment by ccrs** [comment]
add also the enum
````
enum LandmineKnockbackAchievement
{
    SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT  = 57064
};
````

**Comment by tkrokli** [comment]
Yes, that is also a good point. I thought about that only after I posted my test results.
Any advice regarding codestyle / indents compared to the old script by Vincent-Michael from Apr 2012 ?

**Comment by tkrokli** [comment]
Here is the diff after I applied the latest changes, based on the 3.3.5 commit 80ed03a :
```sql
--
-- Landmine Knockback Achievement (29397,54355,54402,57099)
UPDATE `creature_template` SET `flags_extra`= 130 WHERE `entry`= 29397;

DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = 54355;
INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `comment`) VALUES
(54355, 54402, 'Landmine Knockback Achievement');

DELETE FROM `spell_script_names` WHERE `ScriptName` = 'spell_gen_landmine_knockback_achievement';
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
(57099, 'spell_gen_landmine_knockback_achievement');
```
```diff
diff --git a/src/server/scripts/Spells/spell_generic.cpp b/src/server/scripts/Spells/spell_generic.cpp
index ca8c9a9..c690896 100644
--- a/src/server/scripts/Spells/spell_generic.cpp
+++ b/src/server/scripts/Spells/spell_generic.cpp
@@ -4120,6 +4120,44 @@ public:
     }
 };
 
+enum LandmineKnockbackAchievement
+{
+    SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT  = 57064
+};
+
+class spell_gen_landmine_knockback_achievement : public SpellScriptLoader
+{
+    public:
+        spell_gen_landmine_knockback_achievement() : SpellScriptLoader("spell_gen_landmine_knockback_achievement") { }
+
+        class spell_gen_landmine_knockback_achievement_SpellScript : public SpellScript
+        {
+            PrepareSpellScript(spell_gen_landmine_knockback_achievement_SpellScript);
+
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (Player* target = GetHitPlayer())
+                {
+                    Aura const* aura = GetHitAura();
+                    if (!aura || aura->GetStackAmount() != 10)
+                        return;
+
+                    target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
+                }
+            }
+
+            void Register() override
+            {
+                OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_achievement_SpellScript::HandleScript, EFFECT_1, SPELL_EFFECT_APPLY_AURA);
+            }
+        };
+
+    SpellScript* GetSpellScript() const override
+    {
+        return new spell_gen_landmine_knockback_achievement_SpellScript();
+    }
+};
+
 void AddSC_generic_spell_scripts()
 {
     new spell_gen_absorb0_hitlimit1();
@@ -4205,4 +4243,5 @@ void AddSC_generic_spell_scripts()
     new spell_gen_gm_freeze();
     new spell_gen_stand();
     new spell_gen_mixology_bonus();
+    new spell_gen_landmine_knockback_achievement();
 }
```
I tested the suggested changes, but I could not make the achievement work at all. I completely reloaded the DBs with new and empty content in `auth` + `characters` and started with only TDB 335.60 in the `world` DB, so the updates loaded during Worldserver startup. I also disabled all game client addons and deleted both the Cache and WTF folders to be sure there should be nothing to disturb the test. I even made 4 macros for my character, using them 10x in a row each:
- .cast 54355 (Detonation)
- .cast 54402 (Land Mine Knockback)
- .cast 57064 (Land Mine Knockback Achievement)
- .cast 57099 (Land Mine Knockback Achievement Aura)

I also tried replacing 57099 in `spell_script_names` with 57064, since the spellscript `spell_gen_landmine_knockback_achievement` uses 57064 and not 57099 in the .cpp, but then worldserver gave an error saying:
```
Spell `57064` Effect `Index: EFFECT_1 Name: 6` of script
 `spell_gen_landmine_knockback_achievement` did not match dbc effect data
 - handler bound to hook `OnEffectHitTarget` of SpellScript won't be executed
```

But nothing worked. It seems that only Re3os's lines will work in this case. I have modified it slightly:
```diff
diff --git a/src/server/scripts/Spells/spell_generic.cpp b/src/server/scripts/Spells/spell_generic.cpp
index c690896..dc7e582 100644
--- a/src/server/scripts/Spells/spell_generic.cpp
+++ b/src/server/scripts/Spells/spell_generic.cpp
@@ -4122,6 +4122,7 @@ public:
 
 enum LandmineKnockbackAchievement
 {
-    SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT  = 57064
+    LANDMINE_KNOCKBACK_ACHIEVEMENT = 1428
 };
 
@@ -4142,7 +4143,8 @@ class spell_gen_landmine_knockback_achievement : public SpellScriptLoader
                     if (!aura || aura->GetStackAmount() != 10)
                         return;
 
-                    target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
+                    AchievementEntry const* achievement = sAchievementStore.LookupEntry(LANDMINE_KNOCKBACK_ACHIEVEMENT);
+                    target->CompletedAchievement(achievement);
                 }
             }
 
```

![wowscrnshot_121015_092912](https://cloud.githubusercontent.com/assets/4903600/11710617/e0612902-9f20-11e5-959d-a35086d3965d.jpg)

Out of pure curiosity, I searched for similar code in the TC source, and found that a very similar way of handling achievements is done in https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Achievements/AchievementMgr.cpp#L2953 :
```cpp
        if (AchievementEntry const* achievement = sAchievementStore.LookupEntry(i))
```
This code exists in the 6.x AchievementMgr.cpp, but not in 3.3.5, so I am wondering if this is something 3.3.5 lacks and should have had in AchievementMgr.cpp . Anyway, I just wanted to find something validating the use of code that seems to work as intended.

Another theory I have got, but don't know all the validity of, is that maybe the code line
```cpp
target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
```
is missing some parameters, like those used in the `turkey_marker_AuraScript`:
```cpp
target->CastSpell(target, SPELL_TURKEY_VENGEANCE, true, NULL, aurEff, GetCasterGUID());
```
I don't know, I'm just trying to think if something can be done to make the line `target->CastSpell` work.

**Comment by tkrokli** [comment]
OK, so it turns out that the `sAchievementStore.LookupEntry` is illegal/invalid to use except in scripts except for in GM commands. Need to continue researching how the flags work on the Land Mine Bunny.

**Comment by Shauren** [comment]
@tkrokli wrong. Player::CompletedAchievement is the one not allowed to be used.

**Comment by tkrokli** [comment]
Oh. You mean that is the same as `target->CompletedAchievement(achievement);` in this case. Thanks for setting me straight, I am clearly not familiar enough with the code.

**Comment by DevRival** [comment]
[Test](https://gist.github.com/DevRival/1c362a4742c6cde3b07d), work for me.

**Comment by tkrokli** [comment]
@DevRival : Which part? Do you mean the diff where I wrote that I could not make it work?
Or the last change, with the `target->CompletedAchievement(achievement);` script part from Re3os?

**Comment by Re3os** [comment]
@tkrokli sori,My bad english does not allow me to understand many.
p.s I will rewrite  it as a many time

**Comment by DevRival** [comment]
the last change, with the target->CompletedAchievement(achievement);

**Comment by tkrokli** [comment]
OK, that one has been dismissed by Shauren as invalid and only available to GM commands.
If you read the PR comment from MitchesD in https://github.com/TrinityCore/TrinityCore/pull/15977#issuecomment-163572987, he says it can't be done because there is one core issue which must be solved before it can be scripted properly.
(Read the comments in the PR to understand more.)

**Comment by tkrokli** [comment]
TrinityCore rev. 2937af2a5990 2015-12-13 01:14:08 +0100 (landmine_knockback_achievement branch) (Win64, Release) - confirmed. It actually works. I would not have thought that by just looking at the code alone. But it really works. - and yes, like @sirikfoll said, spell_linked already exists in TDB 335.60:
```sql
SELECT * FROM `spell_linked_spell` WHERE `spell_trigger` = 54355;
```
```
spell_trigger  spell_effect    type  comment
-------------  ------------  ------  -----------
        54355         54402       0  Land Mine
```
***

The remaining issue: we need to update table `achievement_criteria_data` too. From DBErrors.log:
```
Table `achievement_criteria_data` does not have expected data for criteria
(Entry: 5258 Type: 29) for achievement 1428.
```
```sql
SELECT * FROM `achievement_criteria_data` WHERE `criteria_id`= 5258 OR `type`= 29;
SELECT * FROM `achievement_criteria_data` WHERE `criteria_id`= 1428 OR `type`= 29;
```
```
criteria_id    type  value1  value2  ScriptName  
-----------  ------  ------  ------  ------------
```
https://trinitycore.atlassian.net/wiki/display/tc/achievement_criteria_data

**Comment by tkrokli** [comment]
@Re3os : You were right after all. Your change in the `disables` table was correct to begin with.
This is the SQL that works and also fixes the DBError issue that I got when using the 5258 update:
```sql
-- Enable Landmine Knockback Achievement
DELETE FROM `disables` WHERE `sourceType`= 4 AND `entry`= 5258;

-- Insert Landmine Knockback Achievement criteria data
DELETE FROM `achievement_criteria_data` WHERE `criteria_id`= 5258;
INSERT INTO `achievement_criteria_data` (`criteria_id`, `type`, `value1`, `value2`, `ScriptName`) VALUES
(5258,0,0,0,'');

-- Insert spell script name for spell 57099
DELETE FROM `spell_script_names` WHERE `ScriptName`='spell_gen_landmine_knockback_achievement';
INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
(57099,'spell_gen_landmine_knockback_achievement');
```
As you can see, the `spell_linked_spell` part was unneeded, because it already exists in TDB 335.60.

**Comment by tkrokli** [comment]
@sirikfoll : I have been thinking about the info inserted into `achievement_criteria_data`,
even if it seems to work with the zeros in the SQL based on your suggestion.

When looking at the table https://trinitycore.atlassian.net/wiki/display/tc/achievement_criteria_data ,
I wonder if maybe that row can use `type` = 5 (`TYPE_S_AURA`), `value1` (`aura spell ID`) = 57099 and `value2` (Effect Index of the aura), but I am uncertain if the "Effect index" = `10` or `0`. Any thoughts?

Maybe I am nitpicking, but I was also thinking about updating the comment in `spell_linked_spell` :
```sql
UPDATE `spell_linked_spell` SET `comment`= 'Trigger Detonation with Land Mine Knockback' WHERE `spell_trigger` = 54355;
```
This idea is based on 2 other spells I have seen linked in the same table:
```
spell_trigger  spell_effect    type  comment
-------------  ------------  ------  ---------------------------------------------------------------------------------
        24023         12021       1  Charge (24023) will now trigger Fixate
        45458         45437       1  Holiday - Midsummer, Stamp Out Bonfire, Quest Complete triggers Stamp Out Bonfire
        52214         51407       2  Looking for treasure aura triggered by Rainspeaker Treasures: See Invisibility
        57469          7448       1  Spread Fire triggers Fire
        31606         31746       0  Stormcrow Amulet triggers Stormcrow Shape
        47190         47189       1  Toaluus Spiritual Incense triggers Spiritual Insight
        54355         54402       0  Trigger Detonation with Land Mine Knockback
        52578         52580       1  Trigger Meat Explosion with Bloody Meat Explosion
        52565         52575       1  Trigger Troll Explosion with FFTF explosion
```

**Comment by sirikfoll** [comment]
Regarding `achievement_criteria_data`
I suggested zero values ​​as a placeholder, the correct values ​​are in the dbc. 
Need some dev to confirm how it's supposed to be

**Comment by tkrokli** [comment]
This is what I could find in the `Achievement_Criteria.dbc` file:
```
 id   field1   field2   field3   field4         field9         field25    field26   field30 
---- -------- -------- -------- -------- -------------------- ---------- --------- ---------
5258   1428      29      57064      1     Landmine Knockback   16712190       2        1    
```
(all the rest of the fields only contained `0` and nothing else. 30 fields in total.)
